lib/sampler/storage/src/lib/span-dao.js

```diff
@@ -0,0 +1,216 @@
+'use strict';
+
+const moment = require('abacus-moment');
+
+const MONGO_DUPLICATE_KEY_ERROR_CODE = 11000;
+
+class SpanDAO {
+  constructor(collection) {
+    this.collection = collection;
+  }
+
+  async startSpan(timestamp, target, measures, dedupID = undefined) {
+    const span = {
+      target: target,
+      measured_usage: measures,
+      start: timestamp,
+      end: null,
+      processing: {
+        complete: false,
+        last_interval: {
+          start: timestamp,
+          end: timestamp
+        },
+        planned_interval: null,
+        last_change_at: moment.now(),
+        version: 1
+      }
+    };
+    if (dedupID)
+      span.start_dedup_id = dedupID;
+    try {
+      await this.collection.insertOne(span);
+      return true;
+    } catch (e) {
+      if (e.code !== MONGO_DUPLICATE_KEY_ERROR_CODE)
+        throw e;
+      return false;
+    }
+  }
+
+  async endSpan(timestamp, target, newCorrelationID, dedupID = undefined) {
+    const filter = {
+      'target': target,
+      'end': null
+    };
+    const update = {
+      '$set': {
+        'target.correlation_id': newCorrelationID,
+        'end': timestamp
+      }
+    };
+    if (dedupID)
+      update.$set.end_dedup_id = dedupID;
+    const options = {};
+    try {
+      const status = await this.collection.findOneAndUpdate(filter, update, options);
+      return !!status.value;
+    } catch (e) {
+      if (e.code !== MONGO_DUPLICATE_KEY_ERROR_CODE)
+        throw e;
+      return false;
+    }
+  }
+
+  async deleteSpansByIDs(ids) {
+    await this.collection.deleteMany({
+      '_id': {
+        '$in': ids
+      }
+    });
+  }
+
+  async getSpanByTarget(target) {
+    const cursor = this.collection.find({
+      'target': target
+    }).limit(1);
+    if (!await cursor.hasNext())
+      return undefined;
+    return await cursor.next();
+  }
+
+  async existsSpanWithStartDedupID(id) {
+    const cursor = this.collection.find({
+      'start_dedup_id': id
+    }).limit(1);
```

Comment:
What's the purpose of hasNext below? If it is to find the next document, then existsSpan is a bit misleading.

Can we simply use findOne instead?

Reply:
The `find` function returns a cursor (as is the variable named). The `hasNext` method of the cursor, as per the driver spec, returns whether there is a next entry. Basically, we are checking whether the find expression matches at all any documents.

Though `findOne` is more explicit on the fact that we need one document (avoiding the `limit` clause), the cursor's `hasNext` might be faster, as it does not need to stream any documents, and I find more to the point.

Others should weigh in on this.

Reply:
We discussed this live and agreed to leave it as is for now.